Docs: add explanation of modules in source directory (#2787)
deepmodeling/abacus-develop: docs/CONTRIBUTING.md

- `tests`: End-to-end test cases;
- `tools`: the script for generating the numerical atomic orbitals.

For those who are interested in the source code, the following figure shows the structure of the source code.

```
|-- module_base                 A basic module including 
|   |                           (1) Mathematical library interface functions: BLAS, LAPACK, Scalapack;
|   |                           (2) Custom data classes: matrix, vector definitions and related functions;
|   |                           (3) Parallelization functions: MPI, OpenMP;
|   |                           (4) Utility functions: timer, random number generator, etc.
|   |                           (5) Global parameters: input parameters, element names, mathematical and physical constants.
|   |-- module_container        The container module for storing data and performing operations on them and on different architectures.
|-- module_basis                Basis means the basis set to expand the wave function.
|   |-- module_ao               Atomic orbital basis set to be refactored.
|   |-- module_nao              New numerical atomic orbital basis set for two-center integrals in LCAO calculations

> |   `-- module_pw               Plane wave basis set for PW calculations

Reviewer: for presently I spend most of time on this module, a more appropriate description might be "For plane wave involved calculations", because in esolver_fp class, we already have PW_Basis* pw_rho to calculate electronic density, and it is, really called in both esolver_ks_pw and esolver_ks_lcao.

> |   `-- module_pw               Plane wave basis set for PW calculations

Reviewer: ```suggestion
|   `-- module_pw               Data structures and relevant methods for planewave involved calculations
```

|-- module_cell                 The module for defining the unit cell and its operations, and reading pseudopotentials.
|   |-- module_neighbor         The module for finding the neighbors of each atom in the unit cell.
|   |-- module_paw              The module for performing PAW calculations.
|   |-- module_symmetry         The module for finding the symmetry operations of the unit cell.
|-- module_elecstate            The module for defining the electronic state and its operations.
|   |-- module_charge           The module for calculating the charge density, charge mixing
|   |-- potentials              The module for calculating the potentials, including Hartree, exchange-correlation, local pseudopotential, etc.

> |-- module_esolver              The module for solving the eigenvalue problem with different theories including Kohn-Sham DFT, Stochastic DFT,

Reviewer: to my best present knowledge and my understanding on our codes, the esolver actually acts as a driver, for different esolver_type defined during parsing of INPUT, corresponding workflow will be assigned and carried out. I have noticed in description of hsolver is already "solving the Hamiltonian with different diagonalization methods, including ...", so my suggestion would be changing present description to "the module defining task-specific driver of corresponding workflow, including lj, dp, ks, sdft, etc."

> |-- module_esolver              The module for solving the eigenvalue problem with different theories including Kohn-Sham DFT, Stochastic DFT,

Reviewer: ```suggestion
|-- module_esolver              The module defining task-specific driver of corresponding workflow for evaluating energies, forces, etc., including lj, dp, ks, sdft, ofdft, etc.
```

|   |-- module_deepks           The module for defining the Hamiltonian in DeepKS calculations.
|   |-- module_dftu             The module for defining the Hamiltonian in DFT+U calculations.
|   |-- module_gint             The module for performing grid integral in LCAO calculations.
|   |-- module_hcontainer       The module for storing the Hamiltonian matrix in LCAO calculations.
|   `-- module_tddft            The module for defining the Hamiltonian in TDDFT calculations.
|-- module_hamilt_pw            The module for defining the Hamiltonian in PW calculations.
|   |-- hamilt_ofdft            The module for defining the Hamiltonian in OFDFT calculations.
|   |-- hamilt_pwdft            The module for defining the Hamiltonian in PW-DFT calculations.
|   |   |-- operator_pw         The module for defining the operators in PW-DFT calculations.
|   `-- hamilt_stodft           The module for defining the Hamiltonian in STODFT calculations.
|-- module_hsolver              The module for solving the Hamiltonian with different diagonalization methods, including CG, Davidson in PW 
|   |                           calculations, and scalapack and genelpa in LCAO calculations.

> |-- module_io                   The module for output properties including band structure, density of states, charge density, etc.

Reviewer: module_io also reads INPUT, INPUTw files. So it is actually what should be refactorized in the future: now that it is named as "io", the reading of KPT file and STRU, orb and upf file are implemented in UnitCell class which is in module_cell module. So temporarily I would suggest add "reading of INPUT files" ahead of the present description.

Author: Thanks for you valuable suggestions. I will add the description!

> |-- module_io                   The module for output properties including band structure, density of states, charge density, etc.

Author: ```suggestion
|-- module_io                   The module for reading of INPUT files and output properties including band structure, density of states, charge density, etc.
```

> |-- module_io                   The module for output properties including band structure, density of states, charge density, etc.

Reviewer: ```suggestion
|-- module_io                   The module for INPUT files read-in and parse, output properties including band structure, density of states, charge density, etc.
```
